### PR TITLE
coreos-copy-firstboot-network: support /etc/coreos-firstboot-network

### LINKED ISSUE #841: Support firstboot networking configuration in `/etc/coreos-firstboot-network`
Right now we support embedding networking configuration into `/boot/coreos-firstboot-network` and having that take effect in the initrd; done to support `coreos-installer install --copy-network`. This works for cases where you have a `/boot/` partition and can leverage it, but doesn't for cases like the Live ISO or Live PXE (though probably less needed for PXE since you probably have DHCP there).

If we additionally supported `/etc/coreos-firstboot-network` then users could layer in files into the ISO initrd and have them applied in the same way as the `/boot/coreos-firstboot-network` does today.

This additional functionality might help people who are trying to build on top of the work that we're doing with special use cases in mind. [BZ#1959961](https://bugzilla.redhat.com/show_bug.cgi?id=1959961#) is the inspiration here.

### PR DESCRIPTION
**Pull request by jlebon**
This adds support for copying NetworkManager keyfiles from
`/etc/coreos-firstboot-network`. It's similar in functionality to the
existing support for `/boot/coreos-firstboot-network`, but is intended
for the live path (ISO or PXE boot) where there is no boot partition.

Users won't interact directly with this; instead, it's expected that
`coreos-installer` will provide high-level commands to customize the ISO
or PXE artifacts to embed keyfiles via the existing layered initrd
mechanisms.

Closes: coreos/fedora-coreos-tracker#841

### CODE REVIEW

**Comment by jlebon** [comment]
I still need to verify that the `/boot/coreos-firstboot-network` path hasn't broken. (We really need tests for this.)

I did verify it works in the ISO path working on top of https://github.com/coreos/coreos-installer/pull/713:

```
$ coreos-installer iso network embed --keyfile ens6.nmconnection builds/latest/x86_64/*.iso --output my.iso
$ cosa run --qemu-iso my.iso -c
...
[root@cosa-devsh ~]# journalctl -u coreos-copy-firstboot-network --no-pager
-- Journal begins at Mon 2021-12-13 22:06:05 UTC, ends at Mon 2021-12-13 22:21:04 UTC. --
Dec 13 22:06:06 localhost systemd[1]: Starting Copy CoreOS Firstboot Networking Config...
Dec 13 22:06:06 localhost coreos-copy-firstboot-network[553]: info: copying files from /etc/coreos-firstboot-network to /run/NetworkManager/system-connections/
Dec 13 22:06:06 localhost coreos-copy-firstboot-network[559]: '/etc/coreos-firstboot-network/ens6.nmconnection' -> '/run/NetworkManager/system-connections/ens6.nmconnection'
Dec 13 22:06:06 localhost systemd[1]: Finished Copy CoreOS Firstboot Networking Config.
Dec 13 22:06:47 localhost systemd[1]: coreos-copy-firstboot-network.service: Deactivated successfully.
Dec 13 22:06:47 localhost systemd[1]: Stopped Copy CoreOS Firstboot Networking Config.

[root@cosa-devsh ~]# journalctl -u coreos-teardown-initramfs --no-pager
-- Journal begins at Mon 2021-12-13 22:06:05 UTC, ends at Mon 2021-12-13 22:21:04 UTC. --
Dec 13 22:06:47 localhost systemd[1]: Stopping CoreOS Tear Down Initramfs...
Dec 13 22:06:47 localhost coreos-teardown-initramfs[830]: info: taking down network device: ens5
Dec 13 22:06:47 localhost coreos-teardown-initramfs[842]: RTNETLINK answers: Operation not supported
Dec 13 22:06:47 localhost coreos-teardown-initramfs[830]: info: flushing all routing
Dec 13 22:06:47 localhost coreos-teardown-initramfs[830]: info: no initramfs hostname information to propagate
Dec 13 22:06:47 localhost coreos-teardown-initramfs[830]: info: no networking config is defined in the real root
Dec 13 22:06:47 localhost coreos-teardown-initramfs[866]: Only in /run/coreos-teardown-initramfs/connections-compare-2/: default_connection.nmconnection
Dec 13 22:06:47 localhost coreos-teardown-initramfs[866]: Only in /run/coreos-teardown-initramfs/connections-compare-1/: ens6.nmconnection
Dec 13 22:06:47 localhost coreos-teardown-initramfs[830]: info: propagating initramfs networking config to the real root
Dec 13 22:06:47 localhost coreos-teardown-initramfs[868]: '/run/NetworkManager/system-connections/ens6.nmconnection' -> '/sysroot/etc/NetworkManager/system-connections/ens6.nmconnection'
Dec 13 22:06:48 localhost coreos-teardown-initramfs[870]: Relabeled /sysroot//etc/NetworkManager/system-connections/ens6.nmconnection from (null) to system_u:object_r:NetworkManager_etc_rw_t:s0
Dec 13 22:06:48 localhost systemd[1]: coreos-teardown-initramfs.service: Deactivated successfully.
Dec 13 22:06:48 localhost systemd[1]: Stopped CoreOS Tear Down Initramfs.

[root@cosa-devsh ~]# file /etc/NetworkManager/system-connections/ens6.nmconnection
/etc/NetworkManager/system-connections/ens6.nmconnection: ASCII text
```

File: overlay.d/05core/usr/lib/dracut/modules.d/35coreos-network/coreos-copy-firstboot-network.sh [hunk 1]
```diff
@@ -4,28 +4,46 @@ set -euo pipefail
...
 
 bootmnt=/mnt/boot_partition
-mkdir -p ${bootmnt}
 bootdev=/dev/disk/by-label/boot
 firstboot_network_dir_basename="coreos-firstboot-network"
-initramfs_firstboot_network_dir="${bootmnt}/${firstboot_network_dir_basename}"
+boot_firstboot_network_dir="${bootmnt}/${firstboot_network_dir_basename}"
+etc_firstboot_network_dir="/etc/${firstboot_network_dir_basename}"
 initramfs_network_dir="/run/NetworkManager/system-connections/"
-realroot_firstboot_network_dir="/boot/${firstboot_network_dir_basename}"
```
Comment (dustymabe) [inline overlay.d/05core/usr/lib/dracut/modules.d/35coreos-network/coreos-copy-firstboot-network.sh:12]: want to add

```
fixup for 277f841
```

to the commit message for this?

File: overlay.d/05core/usr/lib/dracut/modules.d/35coreos-network/coreos-copy-firstboot-network.service [hunk 2]
```diff
@@ -25,21 +25,21 @@
 # [2] https://github.com/dracutdevs/dracut/blob/master/modules.d/35network-manager/nm-config.sh
 # [3] https://github.com/dracutdevs/dracut/blob/master/modules.d/35network-manager/module-setup.sh#L34
 #
+# Recently, this unit also supports running in the live ISO/PXE context, with
+# the NM config specified in `/etc/coreos-firstboot-network` via a layered
+# initrd. See https://github.com/coreos/fedora-coreos-tracker/issues/841.
```
Comment (dustymabe) [inline overlay.d/05core/usr/lib/dracut/modules.d/35coreos-network/coreos-copy-firstboot-network.service]: This comment seems a bit "tacked on". Is there any way that you can see to incorporate it into the original comment already at the top of this script?

**Comment by jlebon** [comment]
Tested this now with https://github.com/coreos/coreos-assembler/pull/2614!

```
$ kola testiso -S --scenarios iso-install --add-nm-keyfile
Testing scenarios: [iso-install]
Ignoring verification of signature on metal image
Successfully tested scenario iso-install for 35.20211213.dev.3 on bios (metal with NM keyfile)
```

File: overlay.d/05core/usr/lib/dracut/modules.d/35coreos-network/coreos-copy-firstboot-network.service [hunk 3]
```diff
@@ -25,21 +25,21 @@
...
 [Unit]
 Description=Copy CoreOS Firstboot Networking Config
 ConditionPathExists=/usr/lib/initrd-release
 DefaultDependencies=false
-Before=ignition-diskful.target
+Before=ignition-complete.target
```
Comment (dustymabe) [inline overlay.d/05core/usr/lib/dracut/modules.d/35coreos-network/coreos-copy-firstboot-network.service:39]: I always get confused with these even though I was probably the one to add `ignition-diskful.target` to begin with. Could we explain why we need this ordering dep specified here (in a comment)?

**Review by dustymabe — DISMISSED** [review]
Mostly LGTM. Some small requests for change.

File: overlay.d/05core/usr/lib/dracut/modules.d/35coreos-network/coreos-copy-firstboot-network.sh [hunk 4]
```diff
@@ -4,28 +4,46 @@ set -euo pipefail
...
 else
-    echo "info: no files to copy from ${initramfs_firstboot_network_dir}. skipping"
+    if [ -n "$(ls -A ${etc_firstboot_network_dir} 2>/dev/null)" ]; then
+        # Also placed there by coreos-installer but in a different flow, see:
+        # https://github.com/coreos/coreos-installer/pull/713
+        copy_firstboot_network "${etc_firstboot_network_dir}"
+    else
+        echo "info: no files to copy from ${etc_firstboot_network_dir}; skipping"
+    fi
 fi
```
Comment (dustymabe) [inline overlay.d/05core/usr/lib/dracut/modules.d/35coreos-network/coreos-copy-firstboot-network.sh:49]: I think this distinction is important. Basically if you are on a live image your firstboot network location is in `/etc/` (in the initramfs). If you are not on a live image your firstboot network location is in `/boot`. There is no mix and match of those and we should keep it that way.

File: overlay.d/05core/usr/lib/dracut/modules.d/35coreos-network/coreos-copy-firstboot-network.service (hunk 3 above)
Comment (bgilbert) [reply overlay.d/05core/usr/lib/dracut/modules.d/35coreos-network/coreos-copy-firstboot-network.service:39]: This PR isn't making that situation any worse, so if we don't have the explanation handy, let's call it out of scope for now.

**Review by bgilbert — DISMISSED** [review]
Thanks!  :fireworks:

File: overlay.d/05core/usr/lib/dracut/modules.d/35coreos-network/coreos-copy-firstboot-network.service (hunk 2 above)
Comment (jlebon) [reply overlay.d/05core/usr/lib/dracut/modules.d/35coreos-network/coreos-copy-firstboot-network.service]: I reworked the big block instead and updated some stale references.

File: overlay.d/05core/usr/lib/dracut/modules.d/35coreos-network/coreos-copy-firstboot-network.service (hunk 3 above)
Comment (jlebon) [reply overlay.d/05core/usr/lib/dracut/modules.d/35coreos-network/coreos-copy-firstboot-network.service:39]: It's just good practice IMO. I added a comment for this.

**Comment by jlebon** [comment]
Updated!

**Review by bgilbert — APPROVED** [review]

**Review by dustymabe — APPROVED** [review]
LGTM

**Pull request merged**